feat: add support for setter prefix and suffix

Techassi opened the pull request:
This PR adds support to add prefixes and suffixes to setter methods.

```rust
#[derive(TypedBuilder)]
#[builder(field_defaults(setter(prefix = "with_", suffix = "_value")))]
struct Foo {
    x: i32,
    y: u32,
}

let foo = Foo::builder().with_x_value(1).with_y_value(2).build();
```

mo8it commented:
May I ask why? It is just more to type repeatedly :/

Techassi commented:
It is nice to have the option. For example I personally like to prefix my (manually implemented) builder methods with `with_`.

This is no breaking change and users don't have to set these values. If they are not set, the setter method is named like the field (current behavior). So really users can choose if they want to type more.

mo8it commented:
I see your point, but I think that this subjective and adds complexity to the code. The whole crate and all its builder attributes have the purpose to type less.

I think that it is fine for a crate to be opionated to some degree.

But let's see what @idanarye thinks about it.

BTW: Always open an issue and wait for the response of the maintainers before starting with a PR. I was disappointed often enough by rejected PRs :)

idanarye commented:
While I myself don't see the point in using `set_` or `with_` prefixes in a builder that does not grant public access to the internal fields, Typed Builder already allows customizing a lot of things so I don't see a point to get opinionated about this thing in particular. And while it does add complexity being a new feature and all, I think that complexity is very localized.

idanarye commented on typed-builder-macro/src/struct_info.rs:
```diff
@@ -285,13 +287,76 @@ impl<'a> StructInfo<'a> {
         })
     }
 
+    fn method_name(field: &FieldInfo, field_name: &Ident) -> Result<Ident, Error> {
```
Please change it to `setter_method_name`. While as a variable `method_name` makes sense inside the context of `field_impl`, it is pretty vague as a struct level method name. Especially when it's under `StructInfo`.

Speaking of which - why is this not under `FieldInfo`? And `FieldInfo` already has it's `name` inside it - there is no need to pass it twice.

Techassi replied:
Yes, that makes sense. Changed in cc1c5da

idanarye commented on typed-builder-macro/src/struct_info.rs:
```diff
@@ -285,13 +287,76 @@ impl<'a> StructInfo<'a> {
+                syn::Expr::Lit(lit) => match &lit.lit {
+                    syn::Lit::Str(str) => Some(syn::Ident::new(&format!("{}", str.value()), proc_macro2::Span::call_site())),
+                    _ => {
+                        return Err(Error::new_spanned(
```
Wouldn't it make more sense to do this check in `SetterSettings::apply_meta`?

idanarye replied:
Also, this whole thing (converting an `syn::Expr` to a string) should probably be extracted to a function.

idanarye commented:
Please add a test with a keyword identifier for a field. Typed Builder had a problem with those in the past (#47), and I suspect this new feature (especially the prefix) may choke on it as well.

idanarye commented:
And please add an entry to the changelog.

Techassi commented:
> Also, I'm not sure how I feel about the prefix and the postfix automatically adding the underscore.

Removed the underscore in cc1c5da



> And please add an entry to the changelog.

Updated in fbae931

Techassi commented:
> Please add a test with a keyword identifier for a field. Typed Builder had a problem with those in the past (#47), and I suspect this new feature (especially the prefix) may choke on it as well.

Added and fixed in ce48427

idanarye commented on src/lib.rs:
```diff
@@ -148,6 +148,14 @@
 ///     when the setter is called.
+///
+///   - `prefix = "..."` prepends the setter method with the specified prefix. For example, setting
+///     `prefix = "with"` results in setters like `with_x` or `with_y`. This option is combinable
```
The example here should say ``prefix = "with_"``. Similarly for the suffix.

idanarye commented on typed-builder-macro/src/util.rs:
```diff
@@ -85,6 +85,17 @@ pub fn strip_raw_ident_prefix(mut name: String) -> String {
+    let len = name.len();
+    name = strip_raw_ident_prefix(name);
+
+    if name.len() < len {
```
1. 
    ```rust
    (name, name.len() < len)
    ```
2. `strip_raw_ident_prefix` is simple enough that it'd be simpler and more straightforward to duplicate its code here instead of 
    * Alternatively - reverse the roles, so that `strip_raw_ident_prefix` calls `strip_raw_ident_prefix_with_status` and ditches the status.
4. A boolean output "named" `status` is not very descriptive.
5. This entire thing is actually redundant - see my comment in `setter_method_name`.

Techassi replied:
Removed in d7e98fc

idanarye commented on typed-builder-macro/src/field_info.rs:
```diff
@@ -73,6 +76,26 @@ impl<'a> FieldInfo<'a> {
     }
 
+    pub fn setter_method_name(&self) -> Ident {
+        let (name, was_stripped) = strip_raw_ident_prefix_with_status(self.name.to_string());
```
You don't actually need `was_stripped`. You only need to re-add the `r#` if both the prefix and the suffix are `None`, but in that case you can just return `self.name` (`.clone()`)

idanarye commented on CHANGELOG.md:
```diff
@@ -10,6 +10,11 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
   setter for that field now does.
 
+### Added
+- Support for setter method pre- and suffixes `#[builder(field_defaults(setter(prefix = "...", suffix = "...")))]`.
```
Nit - "pre-" is confusing, and does not save much compared to just writing "prefixes".

idanarye commented on typed-builder-macro/src/field_info.rs:
```diff
@@ -73,6 +75,20 @@ impl<'a> FieldInfo<'a> {
+        } else if let Some(suffix) = &self.builder_attr.setter.suffix {
+            Ident::new(&format!("{}{}", name, suffix), Span::call_site())
+        } else {
+            Ident::new_raw(&name, Span::call_site())
```
The point is that here you should return `self.name.clone()`, so that if the field name was `r#`'d the setter method name would be `r#`'d as well.

Techassi replied:
Oh man, sorry. Totally missed that. Will change asap.

Techassi replied:
Changed in c38d686
